src/main/kotlin/de/hysky/skyblocker/skyblock/SlotSwap.kt

```diff
@@ -0,0 +1,182 @@
+				currentScreen = null
+			}
+		}
+	}
+
+	private fun onMouseClick(client: MinecraftClient, screen: InventoryScreen, button: Int): Boolean {
+		if (button != GLFW.GLFW_MOUSE_BUTTON_LEFT || !InputUtil.isKeyPressed(client.window.handle, GLFW.GLFW_KEY_LEFT_SHIFT)) return true
+
+		val entry: IntIntImmutablePair = (screen as HandledScreenAccessor).focusedSlot?.id?.let { id ->
+			slotMap.firstOrNull { pair -> pair.keyInt() == id || pair.valueInt() == id }
+		} ?: return true
+
+		if (isHotbarSlot(entry.keyInt())) {
+			client.interactionManager?.clickSlot(screen.screenHandler.syncId, entry.valueInt(), entry.keyInt() - 36, SlotActionType.SWAP, client.player)
+			return false
+		} else if (isHotbarSlot(entry.valueInt())) {
+			client.interactionManager?.clickSlot(screen.screenHandler.syncId, entry.keyInt(), entry.valueInt() - 36, SlotActionType.SWAP, client.player)
+			return false
+		}
+		return true
+	}
+
+	private fun onKeyPress(screen: InventoryScreen, key: Int, scancode: Int) {
+		if (!configureKeybinding.matchesKey(key, scancode) || slotAtKeyPress != null) return
+		slotAtKeyPress = screen.getFocusedSlotId()
+	}
+
+	private fun onKeyRelease(client: MinecraftClient, screen: InventoryScreen, key: Int, scancode: Int) {
+		when {
+			configureKeybinding.matchesKey(key, scancode) -> {
+				val slotAtKeyRelease = screen.getFocusedSlotId()
+				if (slotAtKeyPress != null && slotAtKeyRelease != null && slotAtKeyPress != slotAtKeyRelease
+					&& (isHotbarSlot(slotAtKeyPress!!) || isHotbarSlot(slotAtKeyRelease))
+				) { // At least one slot has to be in the hotbar
+					addSlotMapping(slotAtKeyPress!!, slotAtKeyRelease)
+					client.player?.sendSkyblockerMessage(Text.translatable("skyblocker.slotSwap.add", slotAtKeyPress!!, slotAtKeyRelease))
+				}
+				slotAtKeyPress = null
+			}
+
+			resetKeybinding.matchesKey(key, scancode) -> {
+				if (shouldResetOnNextKey) {
+					client.player?.sendSkyblockerMessage(Text.translatable("skyblocker.slotSwap.reset"))
+					reset()
+					return
+				}
+				shouldResetOnNextKey = true
+				waitJob = CoroutineUtil.globalJob.launch {
+					delay(WAIT_TIME)
+					shouldResetOnNextKey = false //Resets back to the initial state after 500ms, but if the keybinding is pressed again then this timeout is canceled and the reset happens
+				}
+			}
+		}
+	}
+
+	private fun render(screen: InventoryScreen, drawContext: DrawContext) {
+		//Slot mappings
+		for (entry in slotMap) {
+			renderRectanglesAndLine(drawContext, entry.keyInt(), entry.valueInt(), config.sourceSlotColor.rgb, config.targetSlotColor.rgb)
+		}
+		//Configuring state
+		if (slotAtKeyPress == null) return
+		val currentSlot = screen.getFocusedSlotId()
+		renderRectanglesAndLine(drawContext, slotAtKeyPress!!, currentSlot, CONFIGURING_SOURCE_SLOT_COLOR, CONFIGURING_TARGET_SLOT_COLOR)
+	}
+
+	private fun InventoryScreen.getFocusedSlotId() = (this as HandledScreenAccessor).focusedSlot?.let { slot -> if (isSlotValid(slot.id)) slot.id else null }
```

Comment:
How does this work again? Does this pollute `InventoryScreen` so that `getFocusedSlotId` is availbale to everyone or just this file?

Reply:
Oh looks like you need to import it.

Reply:
> How does this work again? Does this pollute `InventoryScreen` so that `getFocusedSlotId` is availbale to everyone or just this file?

It should be only available within this file since it's private. From what I know, it takes the class it is extending as an argument behind the scenes, so it's essentially a static util method. You static import it like in java.
Note that extension functions can also be added in normal classes (this SlotSwap file is a singleton `object`, so everything in it is essentially static), in that case it's just a normal method and can't be used from outside the class. The method itself has to be static to be used from other files.